### PR TITLE
Batch to_verifiable ledger operations, and fix caching logic

### LINKED ISSUE #11705: Batch ledger accesses when calling `User_command.to_verifiable` on a bundle of commands
In order to convert a user command to its verifiable format, verification keys required for verifying parties commands need to be loaded from the ledger. However, in the current implementation, we will do `n` queries to the ledger for this information rather than `1` since we call `to_verifiable` in a loop. We should update the code so that we can call `to_verifiable` in a way such that we can request all of the verification keys from the ledger in batch.

### PR DESCRIPTION
**Pull request by nholland94**
Closes #11705.

This PR modifies `to_all_verifiable` to use batched ledger operations. In the process of implementing this optimization, I discovered that the existing implementation of `to_all_verifiable` did not uphold the properties that we wanted it to, specifically in the scope of the staged ledger's usage of the function. The existing code was making multiple mistakes, including considering verification keys set by failed commands as valid, and only caching a single verification key for a single account at a time (meaning we would drop intermediate keys when switching between accounts across account updates).

I implemented some new staged ledger tests which failed before the code was modified. Those tests now pass with the updated verification key caching code. The way that verification keys are provided from external ledgers has been reworked so that the necessary verification keys can be loaded in batch.

There are a couple of things to bring to the attention of reviewers:
- There is a remaining TODO regarding an optimization for loading less accounts from the ledger. This optimization could be significant, but requires more code changes which need to be vetted and tested very carefully, otherwise we could see bugs due to missing verification keys. I was leaning towards not doing this optimization for now, since batching by itself should help here, but with how this is written, we will request more accounts than necessary (and more accounts than we used to before these code changes).
- The core algorithm for `to_all_verifiable` is shared code, but is split into 2 separate implementations via a functor. It's not clear to me if this is the best approach here, as the algorithms required for the transaction pool and the staged ledger are fairly different and are both critical. It may be best to separate the implementations entirely at the cost of some code duplication. For now, I worked my fix and optimization into the functor abstraction, but let me know if you think it is best to split this up instead.

### CODE REVIEW

**Comment by nholland94** [comment]
!ci-build-me

**Comment by nholland94** [comment]
!ci-build-me

File: src/lib/staged_ledger/pre_diff_info.ml [hunk 1]
```diff
@@ -281,11 +281,21 @@ let get_individual_info (type c) ~constraint_constants coinbase_parts ~receiver
     get_transaction_data ~constraint_constants coinbase_parts ~receiver
       ~coinbase_amount commands completed_works ~to_user_command
   in
+  Core_kernel.Printf.printf
```
Comment (mrmr1993) [inline src/lib/staged_ledger/pre_diff_info.ml:284]: Delete?

File: src/lib/staged_ledger/pre_diff_info.ml [hunk 2]
```diff
@@ -281,11 +281,21 @@ let get_individual_info (type c) ~constraint_constants coinbase_parts ~receiver
...
+    !"number of coinbases = %d, number of fee transfers = %d\n%!"
+    (List.length coinbase_parts)
+    (List.length fee_transfers) ;
   let internal_commands =
     List.map coinbase_parts ~f:(fun t -> Transaction.Coinbase t)
     @ List.map fee_transfers ~f:(fun t -> Transaction.Fee_transfer t)
   in
   let%map internal_commands_with_statuses =
+    Core_kernel.Printf.printf
```
Comment (mrmr1993) [inline src/lib/staged_ledger/pre_diff_info.ml:293]: Delete?

File: src/lib/staged_ledger/staged_ledger.ml [hunk 3]
```diff
@@ -4765,11 +5022,12 @@ let%test_module "staged ledger tests" =
                     ~vk ~ledger zkapp_account_pk ;
                   let invalid_zkapp_command =
                     Or_error.ok_exn
-                      (Zkapp_command.Valid.to_valid ~status:Applied
+                      (Zkapp_command.Valid.to_valid ~failed:false
                          ~find_vk:
-                           (Zkapp_command.Verifiable.find_vk_via_ledger ~ledger
-                              ~get:Ledger.get
-                              ~location_of_account:Ledger.location_of_account )
+                           (Zkapp_command.Verifiable.load_vk_from_ledger
```
Comment (mrmr1993) [inline src/lib/staged_ledger/staged_ledger.ml:5027]: This would benefit from being a helper function.

**Comment by nholland94** [comment]
!ci-build-me

**Comment by nholland94** [comment]
!nightly-build-me

**Comment by nholland94** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
Could you separate the bug fix and the performance improvement changes?

File: src/lib/mina_base/zkapp_command.ml [hunk 4]
```diff
@@ -1217,81 +1236,139 @@ end = struct
...
   end
 
-  module Any = struct
-    include Make_create_all (Map_cache)
+  (* There are 2 situations in which we are converting commands to their verifiable format:
+       - we are reasoning about the validity of commands when the sequence is not yet known
+       - we are reasoning about the validity of commands when the sequence (and by extension, status) is known
+  *)
+
+  module From_unapplied_sequence = struct
```
Comment (mrmr1993) [inline src/lib/mina_base/zkapp_command.ml:1318]: Nit: The comment above reflects that this isn't a sequence, but more of a collection.

File: src/lib/staged_ledger/staged_ledger.ml [hunk 5]
```diff
@@ -4550,6 +4550,262 @@ let%test_module "staged ledger tests" =
...
+              (* we could predict these based on the init ledger, but best to use the proper API *)
+              let location_a = get_location ledger account_a in
+              let location_b = get_location ledger account_b in
+              let%bind cmds =
+                setup_test ledger
+                  (account_a, keypair_a.private_key, location_a)
+                  (account_b, keypair_b.private_key, location_b)
+              in
+              (*
+            let cmds = List.map cmds ~f:(fun cmd ->
```
Comment (mrmr1993) [inline src/lib/staged_ledger/staged_ledger.ml:4603]: Nit: delete

**Review by mrmr1993 — APPROVED** [review]

**Comment by mrmr1993** [comment]
!approved-for-mainnet

**Pull request merged**